sigstore/_internal/rekor/client.py

```diff
@@ -219,14 +193,89 @@ def post(
             },
         }
 
-        resp: requests.Response = self.session.post(self.url, data=json.dumps(data))
+        resp: requests.Response = self.session.post(self.url, json=data)
         try:
             resp.raise_for_status()
         except requests.HTTPError as http_error:
             raise RekorClientError from http_error
 
         return RekorEntry.from_response(resp.json())
 
+    @property
+    def retrieve(self) -> RekorEntriesRetrieve:
+        return RekorEntriesRetrieve(
+            urljoin(self.url, "retrieve/"), session=self.session
+        )
+
+
+class RekorEntriesRetrieve(Endpoint):
+    def post(
+        self,
+        b64_artifact_signature: str,
+        sha256_artifact_hash: str,
+        b64_cert: str,
+    ) -> Optional[RekorEntry]:
+        """
+        Retrieves an extant Rekor entry, identified by its artifact signature,
+        artifact hash, and signing certificate.
+
+        Returns None if Rekor has no entry corresponding to the signing
+        materials.
+        """
+        data = {
+            "entries": [
+                {
+                    "kind": "hashedrekord",
+                    "apiVersion": "0.0.1",
+                    "spec": {
+                        "signature": {
+                            "content": b64_artifact_signature,
+                            "publicKey": {"content": b64_cert},
+                        },
+                        "data": {
+                            "hash": {
+                                "algorithm": "sha256",
+                                "value": sha256_artifact_hash,
+                            }
+                        },
+                    },
+                }
+            ]
+        }
+
+        resp: requests.Response = self.session.post(self.url, json=data)
+        try:
+            resp.raise_for_status()
+        except requests.HTTPError as http_error:
+            if http_error.response.status_code == 404:
+                return None
+            raise RekorClientError(resp.json()) from http_error
+
+        # Rekor should return a 404 when there are no results rather than
+        # an empty list, but we check for the latter just in case.
+        #
+        # Similarly, we expect exactly one top-level response, since we made
+        # exactly one entry request. The inner entry set may have more than
+        # one result, but anything other than 1 at the top-level indicates
+        # a bug in Rekor.
```

Comment:
I'm a little confused by this comment -- does this mean we're expecting a list at the top level with exactly one object in it? The top level response isn't a JSON object? "one top-level response" doesn't parse for me.

Reply:
The top-level response is a JSON array, with each element corresponding to a "search result" for the query sent to the endpoint. We only send one query, so we only expect one response back.

(I agree the language is confusing; I can reword it to something close to what I just wrote.)